### ISSUE TITLE
heartbeat after key expires

### ISSUE THREAD

**Issue by JamesHutchison** [rq/rq#469]
The solution to #138 seems to have created another issue that only occurs in erroneous circumstances. We were running RQ dashboard on a VM and noted that the worker list would disappear randomly even though the workers were still there and still performing jobs on the queue. I was able to isolate the cause to the expiration of the redis key being used in heartbeats and there being no logic to recreate the key once that happens. 

My workaround is to simply register the birth and set the state, like this:

```
def heartbeat(self, timeout=0):
    timeout = max(timeout, self.default_worker_ttl)
    if not self.connection.expire(self.key, timeout):
        # re-register birth if it finds the heartbeat entry missing
        self.register_birth()
        self.set_state(self._state)
    self.log.debug('Sent heartbeat to prevent worker timeout. '
                   'Next one should arrive within {0} seconds.'.format(timeout))
```

**Comment by selwin** [comment]
By "erroneous" circumstances, did you mean that jobs that don't finish successfully? Jobs should be killed if they run beyond their maximum timeout so in theory, worker keys should never expire before the next heartbeat.

Can you provide a test case to reproduce this reliable?

**Comment by JamesHutchison** [comment]
No, I mean on a VM that is frozen for 60+ seconds resulting in the time-out occuring, then the VM runs again. Or maybe a connection issue that eventually resolves itself (i.e. a maintenance window resulting in the VM losing contact with redis temporarily).

Either way, workers shouldn't be in a state where they are doing work but are invisible on the worker list.

**Comment by selwin** [comment]
Ah I see. Agree with your assessment, mind sending a PR with test? :)

**Comment by lwolf** [comment]
Is there any progress on this issue?
I'm experiencing the same problem, after some circumstances my remote workers disappearing from rq-dashboard, but continue performing jobs.

**Comment by nvie** [comment]
See the section "Considerations for jobs" on http://python-rq.org/docs/ for more info on this.

**Comment by stevenkampen** [comment]
@nvie I don't see anything in that docs section relating to this issue. 

I can confirm that this is still an issue. Unless I'm misunderstanding the code, there is no way for a worker to understand that is has been "timed out" via it's redis key expiring (which can happen).

For me this situation happens as a result of a network outage or other temporary non-communication. When communication is restored, the worker sends a heartbeat and continues as normal, but now the worker is in a weird state where it's not registered in its respective queue keys, but is still reading from the queue and executing work. The worker key is (incorrectly) empty except for the `last_timestamp`.

**Comment by selwin** [comment]
I'm reopening this issue. I saw that you already attempted a fix for this [here](https://github.com/stevenkampen/rq/commit/57da38536a0dada048640c28e086519f2fd28a93). I think it's a good approach, just missing some tests. Please open a PR when it's ready.

**Comment by stevenkampen** [comment]
I will look at the tests when I get a chance. Thanks.

**Comment by ll125498a** [comment]
> I'm reopening this issue. I saw that you already attempted a fix for this [here](https://github.com/stevenkampen/rq/commit/57da38536a0dada048640c28e086519f2fd28a93). I think it's a good approach, just missing some tests. Please open a PR when it's ready.

@selwin Hello, I met this situation now, but this link is invalied can you share this solution ? Thanks.

**Comment by lowercase00** [comment]
Steps to reproduce:

1. Start a worker
2. Send a couple of jobs
3. Stop Redis server
4. Leave the worker process alive (on the terminal), it will keep trying to reconnect
5. Wait X minutes (the amount of time for the worker to expire, which is the TTL, should be around 8 minutes)
6. Restart Redis server
7. Send jobs to the queue

Worker will be alive dequeuing and working, but it won't get picked up by `rq info` - even though the `rq:worker:<worker_name>` key exists, the `rq:workers:` namespace doesn't exist anymore, just as the `rq:queues` doesn't exist.

A few relevant things:
1. Heartbeat runs `hset` right after a `expire` command, but doesn't check if the command was already expired, this may be a challenge if it's running within a pipeline, since we don't have access to the result until the `.execute()` was called outside the `heartbeat`.
2. Unreliable registration logic (or checking the registry logic), since it runs `register` (from the `worker_registration` module), but it won't run again in case something went wrong.

One option that might be an easy (but somewhat unreliable) fix, would be to exit the worker process at the same time the Redis key expires.

**Comment by selwin** [comment]
> Heartbeat runs hset right after a expire command, but doesn't check if the command was already expired, this may be a challenge if it's running within a pipeline, since we don't have access to the result until the .execute() was called outside the heartbeat.

By this, did you mean that the heartbeat command arrives after the worker key has been deleted? If that's the case, a clean solution would be to set the heartbeat key only if the key or field exists, similar to `hsetnx` but it's the other way round. Redis doesn't provide this command natively, but we may be able to use a lua script for this: https://stackoverflow.com/questions/53605567/how-to-do-hset-only-if-key-exists-with-redis

> Worker will be alive dequeuing and working, but it won't get picked up by rq info - even though the rq:worker:<worker_name> key exists, the rq:workers: namespace doesn't exist anymore, just as the rq:queues doesn't exist.

I find this strange as the entry from `rq:workers:` is only added from [worker_registration.register](https://github.com/rq/rq/blob/master/rq/worker_registration.py#L18) and removed from `worker_registration.unregister`. So in this case I think the fix should just be to call `unregister()` when the process exits. When it's reconnecting, the worker should still count as being alive.

**Comment by lowercase00** [comment]
It seems currently the worker never stops trying to reconnect. As in this specific case, there's no connection, so we can't unregister. A couple of things:

1. only sets the `heartbeat` if the key was not expired
2. avoid an infinite retry loop

For (1) the lua script should solve, I tried something here but it didn't work, need to look into why:

```python
script = "if redis.call('exists', KEYS[1]) == 1 then redis.call('hset', KEYS[1], ARGV[1], ARGV[2]) end"
connection.eval(script, 1, self.key, 'last_heartbeat', utcformat(utcnow()))
```

For (2), I think there's a choice to make: the "regular" path would be to have a `connection_max_retries`, and I have that ready to go as a PR, but as the worker timeout is defined by the user, maybe the retries should would happen until the worker times out?

Will look further into this.
